### PR TITLE
Limit concurrency

### PR DESCRIPTION
**Pull request by illusional**
Artificially limit concurrency to mitigate resource deadlocking.

### CODE REVIEW

**Review by MattWellie — APPROVED** [review]

File: batch.py [hunk 1]
```diff
@@ -243,6 +243,15 @@ def remove_sc_outliers(df, outliers=['966_967', '88_88']):
     'Space separated, as one argument (Default: all)',
 )
 @click.option('--genes', default=None)
+@click.option(
+    '--max-gene-concurrency',
+    type=int,
+    default=50,
+    help=(
+        'To avoid deadlocking resources, set this concurrency to limit horizontal scale. '
```
Comment (lgruen) [inline batch.py]:
```suggestion
        'To avoid resource starvation, set this concurrency to limit horizontal scale. '
```

File: batch.py [hunk 2]
```diff
@@ -243,6 +243,15 @@ def remove_sc_outliers(df, outliers=['966_967', '88_88']):
...
+        'Higher numbers have a better walltime, but risk deadlocks (which are expensive)'
```
Comment (lgruen) [inline batch.py]:
```suggestion
        'Higher numbers have a better walltime, but risk jobs that are stuck (which are expensive)'
```

File: batch.py [hunk 3]
```diff
@@ -309,6 +319,24 @@ def crm_pipeline(
     else:
         genes_of_interest = list(gene_dict.keys())
 
+    # Setup MAX concurrency by genes
+    _prev_dependent_jobs: list[hb.job.Job] = []
+    _new_dependent_jobs: list[hb.job.Job] = []
+
+    def manage_concurrency_for_job(job: hb.job.Job):
```
Comment (lgruen) [inline batch.py:325]: A maybe simpler implementation of this is to depend on `jobs[my_index - max_concurrency]`.

Comment (illusional) [reply batch.py:325]: Yeah much better idea, I've actioned this, could you review again?

**Review by lgruen — APPROVED** [review]

**Pull request merged**